[MM-12898] Allow in-line markdown images to open preview window

abdusabri opened the pull request:
#### Summary
Updates the behavior of in-line markdown images to allow for opening the preview window. There are 3 cases as follows:

1. If the image source is unsafe, a link is rendered
2. If the image source is safe, the image will have a pointer cursor & a hover effect, and clicking on it will open the preview window
3. If the image is a link itself, the image will be rendered and clicking on it will open the link in a new browser window (no preview or hover effect)

Here is a screen capture demonstrating the expected behavior.

![inline-markdown-8](https://user-images.githubusercontent.com/25670682/64661033-4bbbb180-d443-11e9-96a1-87a2de489c9d.gif)


#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/9805

abdusabri commented on e2e/cypress/integration/markdown/markdown_image_spec.js:
```diff
@@ -42,4 +42,20 @@ describe('Markdown', () => {
+        cy.postMessageFromFile('markdown/markdown_inline_images_2.md');
+
+        // Note we use the Gigantic timeout to ensure that the large images will load
+        cy.getLastPostId().then((postId) => {
```
On my local machine I needed to set the optional timeout parameter for all the `.then` calls (ex. `.then({timeout: TIMEOUTS.LARGE}, (postId) => {`), without it, tests often failed with a timeout and an error `fn.apply is not a function`. I didn't commit that as I'm not sure if this is only an issue on my local machine.

abdusabri replied:
@saturninoabril You would notice that still after my last commit I've not added `wait(TIMEOUTS.SMALL)` to this test - unlike the recent changes from master for the other 2 -, and the test is consistently succeeding without it.

Knowing that waiting for an arbitrary period of time is an anti-pattern, and specially after the fix of `fn.apply` and introducing `waitUntilPermanentPost`, is waiting really needed? why?

Thanks!

saturninoabril replied:
It's needed at the moment since we don't have extra check that the image is already loaded.  But you're definitely correct that it's an anti-pattern.  Maybe one approach is to add the following?
```javascript
// Add width and height check to verify that img element is actually loaded
cy.get(`#postMessageText_${postId}`).find('img').
    ... // other checks
    and('have.attr', 'alt', 'Github').
    and('have.css', 'height', '38px').
    and('have.css', 'width', '38px');
```
If you agree, above suggestion can be addressed in a separate PR.

abdusabri replied:
Sounds good! Let me do it while we are at it, no need to have a separate PR 🙂

abdusabri commented on components/markdown_image.jsx:
```diff
@@ -6,43 +6,100 @@ import React from 'react';
+                        this.props.className :
+                        `${this.props.className} markdown-inline-img--hover cursor--pointer a11y--active`;
+
+                    const getFileExtentionFromUrl = (url) => url.substring(safeSrc.lastIndexOf('.') + 1);
```
On my local machine when I set the `Enable Image Proxy` setting to `true`, I was getting empty `imageMetadata`, but an extension is required for the preview window to work. I'm not sure if this is only an issue on my local environment. Therefore, I opted fro this simple solution. Don't know if it is worth adding a more robust utility function instead.

abdusabri commented on components/file_info_preview/file_info_preview.jsx:
```diff
@@ -24,7 +24,9 @@ export default class FileInfoPreview extends React.PureComponent {
         }
 
-        infoParts.push(Utils.localizeMessage('file_info_preview.size', 'Size ') + Utils.fileSizeToString(fileInfo.size));
+        if (fileInfo.size) {
```
For `.ico` files, `size` was undefined, that's why I've added this condition

abdusabri commented on sass/layout/_markdown.scss:
```diff
@@ -57,12 +57,18 @@ h6 {
         max-height: 500px;
     }
-    div.markdown-inline-img {
+    div .markdown-inline-img {
```
The style was inactive, not sure if that was intentional. I've fixed/activated it

abdusabri commented on utils/text_formatting.jsx:
```diff
@@ -64,6 +64,16 @@ export function formatText(text, inputOptions) {
         output = Markdown.format(output, options);
+        if (output.includes('class="markdown-inline-img"')) {
+            /*
+            ** remove p tag to allow other divs to be nested,
```
Adding this was necessary as with the old implementation there was console errors about wrong DOM nodes as a `<div>` tag shouldn't be nested in a `<p>` tag. And if I don't replace the `<p>` with a `<div>` react produces a weird unique key error (though there are no lists I could find in all the relevant components in the rendering tree)

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-3639.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

esethna commented:
Thanks @abdusabri! Looking really good just one note,

1. We should try to follow the UI that we use for attachments in terms of the hover state:

![image](https://user-images.githubusercontent.com/14333569/64740236-8bb57f80-d4a9-11e9-9503-f880850bfd75.png)
Attachment image ^

![image](https://user-images.githubusercontent.com/14333569/64740276-ac7dd500-d4a9-11e9-8122-223ee59981a4.png)
In-line image ^

___________

FYI I filed a related bug ticket that I found while testing but it's not related to this PR so won't block merging this: https://mattermost.atlassian.net/browse/MM-18508

saturninoabril commented:
@abdusabri fyi, said PR is now merged.

mattermod commented:
New commit detected. SpinWick upgrade will occur after the build is successful.

mattermod commented:
Mattermost test server updated with git commit `650e854c1c1c63c31fed8dc038711da9ce4137c8`.

Access here: https://mattermost-webapp-pr-3639.test.mattermost.cloud

abdusabri commented:
@esethna while updating the tests, I noticed that I'd removed the `alt` attribute of the images by mistake. Sorry, I've fixed this now.

I would like also to note that attached images don't have an `alt` attribute, I think this should be fixed under a separate ticket.

saturninoabril commented on e2e/cypress/integration/markdown/markdown_image_spec.js:
```diff
@@ -46,13 +47,33 @@ describe('Markdown', () => {
+        cy.getLastPostId().then((postId) => {
+            cy.get(`#postMessageText_${postId}`, {timeout: TIMEOUTS.GIGANTIC}).then((post) => {
+                // # Get the image and simulate a click.
+                cy.wrap(post).find('img.markdown-inline-img').first().click();
```
Please add height/width check just to make sure that the image is loaded before clicking.  (Note that we have case where this img element still renders even if there's an error occurred.)

saturninoabril replied:
Also, you mentioned the 3 cases:
1. If the image source is unsafe, a link is rendered
2. If the image source is safe, the image will have a pointer cursor & a hover effect, and clicking on it will open the preview window
3. If the image is a link itself, the image will be rendered and clicking on it will open the link in a new browser window (no preview or hover effect)

Not blocking but would be nice if we have E2E for those.

abdusabri replied:
Thanks @saturninoabril 

Yep, case `1.` is not covered. Thought unit testing is enough for it, but will add it here as well.

`2.` the test named `opens preview window when image is clicked` covers the interactivity aspect of it, and test `with in-line images 2` mostly covers its static aspect/existence. Maybe it is best to combine them together in one test covering the whole use case?

`3.` the test named `with in-line images 1` covers it except for clicking the link and navigating to its destination, which i think should not be the focus of this test suit. However, happy to do so if you would like to test link clicking as well, let me know. Should be renamed as well, right?

Initially, i tried not to do massive changes to the existing tests, I guess that didn't help much 🙃

saturninoabril replied:
1. If unit test covered, then it's all good.
2. Oh okay, then it's good as is (no need to combine)
3. Yup, I agree with it.  Good as is.

Thanks for looking into it!

saturninoabril commented on e2e/cypress/integration/markdown/markdown_image_spec.js:
```diff
@@ -46,13 +47,33 @@ describe('Markdown', () => {
+            });
+        });
+    });
 });
```
Please add test case based from original issue which is posting `![test image](https://www.mattermost.org/wp-content/uploads/2016/03/logoHorizontal.png)` and verifying expected behavior.

abdusabri replied:
Sure, I can use that one

abdusabri commented:
> In RHS, there's no hover effect.
> In center view, the image seemed to be cut-off by a few pixels in standard mode (good in compact).
> ![Screen Shot 2019-09-16 at 7 14 09 PM](https://user-images.githubusercontent.com/5334504/64955241-4d113200-d8ba-11e9-9c46-0b817976489a.png)

Thanks @saturninoabril , didn't know about this case, will fix it. (Getting to know the product 🙈)

abdusabri commented:
> FYI I filed a related bug ticket that I found while testing but it's not related to this PR so won't block merging this: https://mattermost.atlassian.net/browse/MM-18508

@esethna I'm afraid it is actually a regression issue. Made some progress investigating it, will continue working on it.

@saturninoabril I've already fixed the styles for RHS case, and updated E2E tests. But i think it is best to wait until the above issue is fixed before finally pushing, agree?

mattermod commented:
New commit detected. SpinWick upgrade will occur after the build is successful.

abdusabri commented:
Is there a problem preventing the test server from being updated with the latest commit?

saturninoabril commented on e2e/cypress/integration/markdown/markdown_image_spec.js:
```diff
@@ -24,35 +22,83 @@ describe('Markdown', () => {
                 and('have.attr', 'alt', 'Build Status').
-                and('have.attr', 'src', `${baseUrl}/api/v4/image?url=https%3A%2F%2Ftravis-ci.org%2Fmattermost%2Fplatform.svg%3Fbranch%3Dmaster`);
+                and('have.attr', 'src', `${baseUrl}/api/v4/image?url=https%3A%2F%2Ftravis-ci.org%2Fmattermost%2Fplatform.svg%3Fbranch%3Dmaster`).
+                and('have.css', 'height', '22px').
```
Could you please check this again?  It's failing on me where the actual value is `'23px'`.

mattermod commented:
New commit detected. SpinWick upgrade will occur after the build is successful.

abdusabri commented:
> It's failing on me where the actual value is '23px'.

@saturninoabril, Hmmm 🤔, don't know why there was `1px` difference, it was passing for me with `22px` and `23px` was failing. 

I've made a minor CSS optimization/tweak, which applies to the case in hand (image is a link), and I've adjusted the values accordingly in the E2E test. Hopefully this makes it consistent for you as well.

mattermod commented:
Mattermost test server updated with git commit `96a5fda66276aedf8b35740ee32147880bffdeca`.

Access here: https://mattermost-webapp-pr-3639.test.mattermost.cloud

saturninoabril commented:
@abdusabri Thanks looks good to me.  All tests passed.

abdusabri commented:
@sudheerDev, should I wait for your review before merging and pushing? (to minimize the number of merge commits)

sudheerDev commented:
@abdusabri Its fine either way. You choice. I will do the review today.

sudheerDev commented:
@abdusabri Nice work!. Can you merge master into the branch?

mattermod commented:
Test server destroyed

abdusabri commented:
Thanks for the feedback @sudheerDev!

abdusabri commented:
> I would like also to note that attached images don't have an alt attribute, I think this should be fixed under a separate ticket.

@esethna, in case you missed ☝️
